dora/core/common/src/main/java/alluxio/uri/UfsUrl.java

```diff
@@ -0,0 +1,414 @@
+      } else if (ufsUrlMessage.getAuthority().isEmpty()) {
+        pathComponents.addAll(rootMessage.getPathComponentsList());
+      }
+    }
+
+    pathComponents.addAll(ufsUrlMessage.getPathComponentsList());
+
+    return new UfsUrl(UfsUrlMessage.newBuilder()
+        .setScheme(scheme)
+        .setAuthority(authority)
+        .addAllPathComponents(pathComponents)
+        .build());
+  }
+
+  /**
+   * Construct an UfsUrl from an UfsUrlMessage.
+   * @param proto the proto message
+   * @return an UfsUrl object
+   */
+  public static UfsUrl fromProto(UfsUrlMessage proto) {
+    return new UfsUrl(proto);
+  }
+
+  /**
+   * Construct an {@link UfsUrl} from components.
+   * @param proto the proto of the UfsUrl
+   */
+  public UfsUrl(UfsUrlMessage proto) {
+    Preconditions.checkArgument(proto.getPathComponentsList().size() != 0,
+        "The proto.path is empty, please check the proto first.");
+    mProto = proto;
+  }
+
+  /**
+   * Construct an {@link UfsUrl} from components.
+   * @param scheme the scheme of the path
+   * @param authority the authority of the path
+   * @param path the path component of the UfsUrl
+   */
+  public UfsUrl(String scheme, String authority, String path) {
+    String[] arrayOfPath = path.split(PATH_SEPARATOR);
+    List<String> pathComponentsList = Arrays.asList(arrayOfPath);
+    mProto = UfsUrlMessage.newBuilder()
+        .setScheme(scheme)
+        .setAuthority(authority)
+        .addAllPathComponents(pathComponentsList)
+        .build();
+  }
+
+  /**
+   * @return the scheme of the {@link UfsUrl}
+   */
+  public Optional<String> getScheme() {
+    if (!mProto.hasScheme()) {
+      return Optional.empty();
+    }
+    return Optional.of(mProto.getScheme());
+  }
+
+  /**
+   * @return the authority of the {@link UfsUrl}
+   */
+  public Optional<Authority> getAuthority() {
+    if (!mProto.hasAuthority()) {
+      return Optional.empty();
+    }
+    return Optional.of(Authority.fromString(mProto.getAuthority()));
+  }
+
+  /**
+   * @return the pathComponents List of the {@link UfsUrl}
+   */
+  // TODO(Tony Sun): In the future Consider whether pathComponents should be extracted as a class.
+  public List<String> getPathComponents() {
+    return mProto.getPathComponentsList();
+  }
+
+  /**
+   * @return the proto field of the {@link UfsUrl}
+   */
+  public UfsUrlMessage getProto() {
+    return mProto;
+  }
+
+  /**
+   * @return the String representation of the {@link UfsUrl}
+   */
+  public String asString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append(mProto.getScheme());
+    if (!mProto.getScheme().isEmpty()) {
+      sb.append(UfsUrl.SCHEME_SEPARATOR);
+    }
+    sb.append(mProto.getAuthority());
+    if (!mProto.getAuthority().isEmpty()) {
+      sb.append(UfsUrl.PATH_SEPARATOR);
+    }
+    List<String> pathComponents = mProto.getPathComponentsList();
+    for (int i = 0; i < pathComponents.size(); i++) {
+      if (pathComponents.get(i).isEmpty())  {
```

Comment:
will there still be empty elements in the list?

Reply:
Yes, in some strange test cases, if I remove the isEmpty judgement, the test fails.